tests/unit_tests/Image_test.cc

```diff
@@ -124,6 +124,32 @@ class ImageTest : public ::testing::Test {
     int size_;
 };
 
+namespace VCL {
+
+    class ImageTest : public Image{
+
```

Comment:
Were any of the tests failing in the first commit that removed ImageData? If yes, either it would make sense to fix it in the same commit or remove them before hand to avoid a non-working commit OR make a note in the commit message.

Reply:
Yes, there are actually build errors due to the presence of private constructors and other functions. Except that, the tests passed fine.

The changes on merging the classes (f3966e6: Merge ImageData and Image classes) and merging the unit-tests (e8cd292: Merge unit ...)  are done separately, so there'll be build errors if f3966e6 is applied stand-alone. Reason for this was to not make changes TestImageData_test class, which would be removed in e8cd292 anyway. 

Fixing the errors on the same commit will result in writing code that'll be removed in the next patch (since some of the tests are removed when merging ImageData and Image test classes, as they're duplicates).

I guess the best would be to squash those three commits into one, but that'd make the change even bigger. Not sure what's the best approach here.

Reply:
I think we have three choices (as Vishakha indicated).
0) What we have now: some tests are fixed in the inital commit, but some other fail.
1) Don't make any changes to the tests in the initial commit; add a note to the commit message that tests will fail until the next commit. This has long-term affects if we ever want to do a git-bisect.
2) Make sure that all the tests pass. I guess this effectively means merging all three commits.
3) Remove the ImageData tests in a commit prior to the commit that removes ImageData. In the commit that merges Image and ImageData, make the minimal changes to the tests so that they still pass. In a subsequent commit, add the new Image tests to test the "new" functions in the Image class.

Reply:
If we go with 1st option, we'll have build errors due to constructors/functions that became private members as result of the merging. Furthermore, as you mentioned, hitting that commit as result of a git-bisect will be unpleasant. Let me take care of the build errors (by applying the 'friend' class trick, just to be removed in next patch). If the tests pass, we should be good. The only downside is I'll write code that's deleted just in the next patch, but I guess it's pretty much the same effort with the 3rd option.

2nd option looks the easiest. I can squash  f3966e6 and e8cd292, and rebase the move constructor change on the squashed changes. The change will be huge, though.

Reply:
Seems like the option 3 might be the cleanest and easiest. That way you are not changing tests that get deleted. Just adding new test cases after the functionality. Is that not true? And none of your commits will break most likely.

Reply:
Agreed, let's do the third way. I've already tested the 1st option, and even though there are no UT errors once the build errors are resolved, the 3rd option seems cleaner as we do not add any lines that are removed just in next patch.

There may be merge conflicts, but shouldn't be much of an issue. I'll apply the 3rd option and push the changes.